pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler_test.go

```diff
@@ -0,0 +1,341 @@
+				return fake.NewMockNonNamespacedControllerInterface[*rbac.ClusterRole, *rbac.ClusterRoleList](ctrl)
+			},
+			gr: &v3.GlobalRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: grName,
+					UID:  grUID,
+				},
+				InheritedFleetWorkspacePermissions: v3.FleetWorkspacePermission{
+					ResourceRules:  resourceRules,
+					WorkspaceVerbs: workspaceVerbs,
+				},
+			},
+		},
+		"backing Roles and ClusterRoles are updated with new content": {
+			crCache: clusterRoleMock(ctrl),
+			crClient: func() rbacv1.ClusterRoleController {
+				mock := fake.NewMockNonNamespacedControllerInterface[*rbac.ClusterRole, *rbac.ClusterRoleList](ctrl)
+				// expect updates with new rules
+				mock.EXPECT().Update(&rbac.ClusterRole{
+					TypeMeta: metav1.TypeMeta{},
+					ObjectMeta: metav1.ObjectMeta{
+						Name: wrangler.SafeConcatName(grName, fleetWorkspaceClusterRulesName),
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "management.cattle.io/v3",
+								Kind:       "GlobalRole",
+								Name:       grName,
+								UID:        grUID,
+							},
+						},
+					},
+					Rules: []rbac.PolicyRule{
+						{
+							Verbs:     []string{"*"},
+							APIGroups: []string{"fleet.cattle.io"},
+							Resources: []string{"gitrepos"},
+						},
+					},
+				})
+				mock.EXPECT().Update(&rbac.ClusterRole{
+					TypeMeta: metav1.TypeMeta{},
+					ObjectMeta: metav1.ObjectMeta{
+						Name: wrangler.SafeConcatName(grName, fleetWorkspaceVerbsName),
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "management.cattle.io/v3",
+								Kind:       "GlobalRole",
+								Name:       grName,
+								UID:        grUID,
+							},
+						},
+					},
+					Rules: []rbac.PolicyRule{
+						{
+							Verbs:         []string{"*"},
+							APIGroups:     []string{"management.cattle.io"},
+							Resources:     []string{"fleetworkspaces"},
+							ResourceNames: []string{"fleet-default"},
+						},
+					},
+				})
+				return mock
+			},
+			gr: &v3.GlobalRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: grName,
+					UID:  grUID,
+				},
+				InheritedFleetWorkspacePermissions: v3.FleetWorkspacePermission{
+					ResourceRules: []rbac.PolicyRule{
+						{
+							Verbs:     []string{"*"},
+							APIGroups: []string{"fleet.cattle.io"},
+							Resources: []string{"gitrepos"},
+						},
+					},
+					WorkspaceVerbs: []string{"*"},
+				},
+			},
+		},
+	}
+
+	for name, test := range tests {
+		test := test
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			h := fleetWorkspaceRoleHandler{
+				crClient: test.crClient(),
+				crCache:  test.crCache(),
+				fwCache:  fleetDefaultAndLocalWorkspaceCacheMock(ctrl),
+			}
+
+			err := h.reconcileFleetWorkspacePermissions(test.gr)
+
+			assert.Equal(t, err, nil)
+		})
+	}
+}
+
+func TestReconcileFleetPermissions_errors(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	tests := map[string]struct {
```

Comment:
Some test cases that would be good to add:
- We could get the first clusterrole, but not the second (for workspace verbs)
- We couldn't update the first/second cluster role

Reply:
I might be missing something but I don't see the first bullet point (an error other than not found when getting the workspace verbs CR) or the second half of the second bullet.

Overall, I would recommend running the go coverage tool and looking to correct most of the missed red lines (`go test -coverprofile=profile.cov` and `go tool cover -html=profile.cov`). You don't have to cover the "new" functions, but you should get most of the other missed lines, particularly the error cases.

Reply:
test added in the last commit